build-tools/packages/build-cli/src/library/changesets.ts

```diff
@@ -125,24 +297,103 @@ export function groupByPackage(
 }
 
 /**
- * Given an array of changesets, flattens the changesets into an array of ChangesetEntry objects.
+ * Creates a map of package names to an array of all the changesets that apply to the package. Only the "main" package
+ * is considered. The returned array of changesets is sorted oldest-to-newest (that is, index 0 is the earliest
+ * changeset, and the last changeset in the array is the newest).
+ *
+ * @param changesets - An array of changesets to be grouped.
+ * @returns a Map of package names to an array of all the changesets that apply to the package.
+ */
+export function groupByMainPackage(changesets: Changeset[]): Map<ReleasePackage, Changeset[]> {
+	const changesetMap = new Map<ReleasePackage, Changeset[]>();
+	for (const changeset of changesets) {
+		const entries = changesetMap.get(changeset.mainPackage) ?? [];
+		entries.push(changeset);
+		changesetMap.set(changeset.mainPackage, entries);
+	}
+
+	// Sort all the entries by highlighted status and date
+	for (const entries of changesetMap.values()) {
+		entries.sort(compareChangesets);
+	}
+
+	return changesetMap;
+}
+
+/**
+ * Creates a map of section names to an array of all the changesets that belong in that section.
+ *
+ * The returned array of changesets are sorted oldest-to-newest (that is, index 0 is the earliest changeset, and the
+ * last changeset in the array is the newest).
+ *
+ * Any changesets that do not belong to a section will be in the {@link UNKNOWN_SECTION} (`_unknown`) key in the
+ * returned map, so callers should check the contents of that key to ensure all changesets were mapped to sections as
+ * expected.
+ *
+ * Note that this groups by the section values in the changesets. Callers are expected to validate tha section names or
+ * adjust them depending on their needs. This function does not adjust section names _except_ for those with no
+ * specified section.
+ *
+ * @param changesets - An array of changesets to be grouped.
+ * @returns a Map of section names to an array of all the changesets that apply to that section.
+ */
+export function groupBySection(
+	changesets: Changeset[],
+): Map<ReleaseNotesSection["name"], Changeset[]> {
```

Comment:
Assuming callers are not meant to mutate this...
```suggestion
): ReadonlyMap<ReleaseNotesSection["name"], readonly Changeset[]> {
```

Reply:
(same with the `groupByMainPackage`)